staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/etcd.go

```diff
@@ -50,5 +87,154 @@ func NewREST(resource schema.GroupResource, listKind schema.GroupVersionKind, st
+var _ = rest.Patcher(&ScaleREST{})
+var _ = rest.GroupVersionKindProvider(&ScaleREST{})
+
+func (r *ScaleREST) GroupVersionKind(containingGV schema.GroupVersion) schema.GroupVersionKind {
+	return autoscalingv1.SchemeGroupVersion.WithKind("Scale")
+}
+
+// New creates a new Scale object
+func (r *ScaleREST) New() runtime.Object {
+	return &autoscalingv1.Scale{}
+}
+
+func (r *ScaleREST) Get(ctx genericapirequest.Context, name string, options *metav1.GetOptions) (runtime.Object, error) {
+	cr, err := r.registry.GetCustomResource(ctx, name, options)
+	if err != nil {
+		return nil, err
+	}
+
+	scaleObject, err := scaleFromCustomResource(cr, r.specReplicasPath, r.statusReplicasPath, r.labelSelectorPath)
+	if err != nil {
+		return nil, err
+	}
+	return scaleObject, err
+}
+
+func (r *ScaleREST) Update(ctx genericapirequest.Context, name string, objInfo rest.UpdatedObjectInfo, createValidation rest.ValidateObjectFunc, updateValidation rest.ValidateObjectUpdateFunc) (runtime.Object, bool, error) {
+	cr, err := r.registry.GetCustomResource(ctx, name, &metav1.GetOptions{})
+	if err != nil {
+		return nil, false, err
+	}
+
+	oldScale, err := scaleFromCustomResource(cr, r.specReplicasPath, r.statusReplicasPath, r.labelSelectorPath)
+	if err != nil {
+		return nil, false, err
+	}
+
+	obj, err := objInfo.UpdatedObject(ctx, oldScale)
+	if err != nil {
+		return nil, false, err
+	}
+	if obj == nil {
+		return nil, false, apierrors.NewBadRequest(fmt.Sprintf("nil update passed to Scale"))
+	}
+
+	scale, ok := obj.(*autoscalingv1.Scale)
+	if !ok {
+		return nil, false, apierrors.NewBadRequest(fmt.Sprintf("wrong object passed to Scale update: %v", obj))
+	}
+
+	specReplicasPath := strings.TrimPrefix(r.specReplicasPath, ".") // ignore leading period
+	if ok = unstructured.SetNestedField(cr.Object, int64(scale.Spec.Replicas), strings.Split(specReplicasPath, ".")...); !ok {
+		return nil, false, fmt.Errorf("spec replicas value cannot be set because one of the nesting levels is not a map[string]interface{}")
+	}
+	cr.SetResourceVersion(scale.ResourceVersion)
+
+	cr, err = r.registry.UpdateCustomResource(ctx, cr, createValidation, updateValidation)
+	if err != nil {
+		return nil, false, err
+	}
+
+	newScale, err := scaleFromCustomResource(cr, r.specReplicasPath, r.statusReplicasPath, r.labelSelectorPath)
+	if err != nil {
+		return nil, false, apierrors.NewBadRequest(fmt.Sprintf("%v", err))
+	}
+	return newScale, false, err
+}
+
+// scaleFromCustomResource returns a scale subresource for a customresource.
+func scaleFromCustomResource(cr *unstructured.Unstructured, specReplicasPath, statusReplicasPath, labelSelectorPath string) (*autoscalingv1.Scale, error) {
+	specReplicasPath = strings.TrimPrefix(specReplicasPath, ".") // ignore leading period
+	specReplicas, found, err := unstructured.NestedInt64(cr.UnstructuredContent(), strings.Split(specReplicasPath, ".")...)
+	if err != nil {
+		return nil, err
+	} else if !found {
+		specReplicas = 1
```

Comment:
@sttts ptal. We should be setting this to 1, instead of 0 if the path is not found. That's what we follow in existing resources: https://github.com/kubernetes/api/blob/master/apps/v1beta2/types.go#L166.

Reply:
Don't like that. But maybe @DirectXMan12 @liggitt @deads2k have another opinion.

Reply:
Just to be clear: this is the (usually non-optional) spec value, not the status (which defaults to zero for CRDs if not found in the CR).

Reply:
> We should be setting this to 1, instead of 0 if the path is not found. That's what we follow in existing resources: https://github.com/kubernetes/api/blob/master/apps/v1beta2/types.go#L166.

that's a decision that API made, not a universal default. we can either on get if the specified field doesn't exist, or let them specify what the default scale is

Reply:
I tend to see a scale object without specReplicas as invalid. I.e. validation on POST/PUT should fail, GET should error (50x). 

IMO we should either follow the old semantics of 1 or error out. Adding a default sounds like avoiding an decission here.

@deads2k @DirectXMan12 @liggitt please comment to find an agreement.

Reply:
The results of another discussion:

- native resources like `StatefulSet`, `ReplicaSet` default their `spec.replicas` to a certain value. But they never persist a `nil` value. Hence, the `/scale` subresource always sees a non-nil `spec.replicas` value.
- Hence, CustomResources with a non-existing `spec.replicas` value are a special case. In order to avoid guessing and another CRD field without clear use-case, we should return an error on `/scale` GET requests in that case.
- The `ScaleSpec.Replicas` value is optional, but defaults to zero. So technically on `/scale` update calls `ScaleSpec.Replicas` is always set (to zero).
- An `/scale` update request against a CustomResource with a non-existing `spec.replicas` value should set the `spec.replicas` value, independenty from whether it was existing or not before the request.

In other words: the defaulting of a CustomResource's `spec.replicas` will not happen magically inside the /scale subresource, but must be done by other mechanisms (admission webhook today; or JSON Schema validation, some day).

Reply:
> In other words: the defaulting of a CustomResource's spec.replicas will not happen magically inside the /scale subresource, but must be done by other mechanisms (admission webhook today; or JSON Schema validation, some day).

Very much agree with this.  Defaulting happens at the base layer, not the subresource "view" layer.

Reply:
Failing on GET seems like the safest and simplest option.  As @DirectXMan12 noted, the responsibility for the validity of the data lies with the core resource, not the subresources.  Return zero instead of failing could produce weird behavior in things that expect a correct value.  Without default values in general, I'm against creating special defaulting field.